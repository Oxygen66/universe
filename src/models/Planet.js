import database from '../../database';

class Planet {
  #db = database;

  async getAll() {
    return this.#db('planet');
  }

  async getBySpaceCenterId(id) {
    return this.#db('planet')
      .select('planet.*')
      .innerJoin(
        'space_center',
        'space_center.planet_id',
        'planet.id',
      )
      .where('space_center.id', id)
      .first();
  }
}

export default Planet;
