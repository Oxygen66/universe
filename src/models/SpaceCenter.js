import database from '../../database';

class SpaceCenter {
  #db = database;

  async getByPlanetIdAndLimit(planetId, limit) {
    return this.#db('space_center')
      .where('planet_id', planetId)
      .limit(limit);
  }

  async getById(id) {
    return this.#db('space_center')
      .where({ id })
      .first();
  }

  async getByUid(uid) {
    return this.#db('space_center')
      .where({ uid })
      .first();
  }

  async getAllPaginated(page, pageSize) {
    return this.#db('space_center')
      .limit(pageSize)
      .offset((pageSize - 1) * page);
  }

  async countAll() {
    return this.#db('space_center').count({ total: 'id' });
  }
}

export default SpaceCenter;
