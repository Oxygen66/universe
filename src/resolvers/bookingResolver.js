import { UserInputError } from 'apollo-server-koa';

const bookingResolver = {
  Mutation: {
    bookFlight: async (
      _,
      { bookingInfo: { seatCount, flightId, email } },
      { dataSources: { bookingAPI, flightAPI } },
    ) => {
      if (seatCount < 1) {
        throw new UserInputError('seatCount should be greater than 0', {
          invalidArgs: 'seatCount',
        });
      }

      const flight = await flightAPI.getById(flightId);

      const bookings = await bookingAPI.getAllByFlightId(flightId);
      const remainingSeats = [...bookings]
        .map((booking) => booking.seat_count)
        .reduce((previousValue, currentValue) => previousValue - currentValue, flight.seat_count);

      if ((remainingSeats - seatCount) < 0) {
        throw new UserInputError('seatCount must be less than the number of remaining seats', {
          invalidArgs: 'seatCount',
        });
      }

      const [bookingResult] = await bookingAPI.createBooking(seatCount, flightId, email);
      return bookingResult;
    },
  },
  Query: {
    booking: (_, { id }, { dataSources: { bookingAPI } }) => bookingAPI.getById(id),
  },
  Booking: {
    id: (parent) => parent.id,
    seatCount: (parent) => parent.seat_count,
    email: (parent) => parent.email,
    flight: (parent, args, { dataSources: { flightAPI } }) => flightAPI.getById(parent.flight_id),
  },
};

export default bookingResolver;
