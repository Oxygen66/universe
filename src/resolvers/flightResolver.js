import { UserInputError } from 'apollo-server-koa';

const flightResolver = {
  Query: {
    flight: (parent, { id }, { dataSources: { flightAPI } }) => flightAPI.getById(id),
    flights:
      async (
        parent,
        {
          page, pageSize, seatCount, from, to, departureDay,
        },
        { dataSources: { flightAPI } },
      ) => {
        if (page < 1) {
          throw new UserInputError('Page must be greater than 1', {
            invalidArgs: 'page',
          });
        }

        if (pageSize < 1 || pageSize > 100) {
          throw new UserInputError('PageSize must be between 1 and 100', {
            invalidArgs: 'page',
          });
        }

        return {
          flights: await flightAPI
            .getAllPaginated(page, pageSize, from, to, seatCount, departureDay),
          page,
          pageSize,
          from,
          to,
          seatCount,
          departureDay,
        };
      },
  },
  Mutation: {
    scheduleFlight:
      async (
        _,
        {
          flightInfo: {
            launchSiteId,
            landingSiteId,
            departureAt,
            seatCount,
          },
        },
        { dataSources: { flightAPI } },
      ) => {
        if (landingSiteId === launchSiteId) {
          throw new UserInputError('landingSiteId and landingSiteId cannot be the same', {
            invalidArgs: [
              'launchSiteId', 'landingSiteId',
            ],
          });
        }

        if (!departureAt) {
          throw new UserInputError('departureAt must be a valid ISO date', {
            invalidArgs: 'departureAt',
          });
        }

        const [flightResult] = await flightAPI
          .createFlight(launchSiteId, landingSiteId, departureAt, seatCount);
        return flightResult;
      },
  },
  Flight: {
    id: (parent) => parent.id,
    code: (parent) => parent.code,
    departureAt: (parent) => parent.departure_at,
    seatCount: (parent) => parent.seat_count,
    launchSite: (
      parent, args, { dataSources: { spaceCenterAPI } },
    ) => spaceCenterAPI.getById(parent.launching_site),
    landingSite: (
      parent, args, { dataSources: { spaceCenterAPI } },
    ) => spaceCenterAPI.getById(parent.landing_site),
    availableSeats: async (
      { id, seat_count: seatCount }, args, { dataSources: { bookingAPI } },
    ) => {
      const bookings = await bookingAPI.getAllByFlightId(id);
      return [...bookings]
        .map((booking) => booking.seat_count)
        .reduce((previousValue, currentValue) => previousValue - currentValue, seatCount);
    },
  },
  FlightsResult: {
    pagination: async (
      {
        page, pageSize, from, to, seatCount, departureDay,
      },
      args,
      { dataSources: { flightAPI } },
    ) => {
      const [{ total }] = await flightAPI.countAll(from, to, seatCount, departureDay);
      return {
        total,
        page,
        pageSize,
      };
    },
    nodes: (parent) => parent.flights,
  },
};

export default flightResolver;
