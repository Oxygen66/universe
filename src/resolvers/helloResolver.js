const helloResolvers = {
  Query: {
    hello: () => 'Hello Adrien',
  },
};

export default helloResolvers;
