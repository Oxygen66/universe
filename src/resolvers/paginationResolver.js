const paginationResolver = {
  Pagination: {
    total: (parent) => parent.total,
    page: (parent) => parent.page,
    pageSize: (parent) => parent.pageSize,
  },
};

export default paginationResolver;
