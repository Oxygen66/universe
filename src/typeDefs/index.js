import types from './types';
import query from './query';
import mutation from './mutation';
import inputs from './inputs';

const typeDefs = [
  ...types,
  ...inputs,
  query,
  mutation,
];

export default typeDefs;
