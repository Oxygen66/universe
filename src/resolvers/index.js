import { merge } from 'lodash';
import planetResolver from './planetResolver';
import spaceCenterResolver from './spaceCenterResolver';
import paginationResolver from './paginationResolver';

const resolvers = merge(
  planetResolver,
  spaceCenterResolver,
  paginationResolver,
);

export default resolvers;
