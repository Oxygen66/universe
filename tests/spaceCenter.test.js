import { ApolloServer, gql } from 'apollo-server-koa';
import { createTestClient } from 'apollo-server-testing';
import Planet from '../src/models/Planet';
import SpaceCenter from '../src/models/SpaceCenter';
import typeDefs from '../src/typeDefs';
import resolvers from '../src/resolvers';

const planetAPI = new Planet();
const spaceCenterAPI = new SpaceCenter();

const server = new ApolloServer({
  typeDefs,
  resolvers,
  dataSources: () => ({ planetAPI, spaceCenterAPI }),
});

const GET_SPACE_CENTERS = gql`
    query spaceCenters {
        spaceCenters {
            pagination {
                total
                page
                pageSize
            }
            nodes {
                id
                uid
                name
                description
                latitude
                longitude
                planet {
                    id
                    name
                    code
                }
            }
        }
    }
`;

const GET_SPACE_CENTER_BY_CUSTOM_VARS = gql`
    query spaceCenters($page: Int!, $pageSize: Int!) {
        spaceCenters(page: $page, pageSize: $pageSize) {
            pagination {
                total
                page
                pageSize
            }
            nodes {
                id
                uid
                name
                description
                latitude
                longitude
                planet {
                    id
                    name
                    code
                }
            }
        }
    }
`;

describe('Queries', () => {
  it('should test', () => {
    expect(true).toBeTruthy();
  });

  describe('spaceCenters', () => {
    beforeEach(() => {
      planetAPI.getBySpaceCenterId = jest.fn(() => ({
        id: 1,
        name: 'test1',
        code: 'COD',
      }));
      spaceCenterAPI.getAllPaginated = jest
        .fn((page, pageSize) => [...Array(pageSize)].map((_, i) => {
          const id = (page - 1) * pageSize + i + 1;
          return ({
            id,
            uid: `uid${id}`,
            name: `name${id}`,
            description: `name${id}`,
            latitude: -1.0,
            longitude: -1.0,
            planet_id: 1,
          });
        }));
      spaceCenterAPI.countAll = jest.fn(() => [{ total: 1081 }]);
    });

    it('fetch space centers ', async () => {
      const { query } = createTestClient(server);
      const res = await query({
        query: GET_SPACE_CENTERS,
      });
      expect(res.errors).not.toBeTruthy();
      expect(res).toMatchSnapshot();
    });

    it('fetch space centers ', async () => {
      const { query } = createTestClient(server);
      const res = await query({
        query: GET_SPACE_CENTERS,
      });
      expect(res.errors).not.toBeTruthy();
      expect(res).toMatchSnapshot();
    });

    it('fetch spacers with custom page, pageSize', async () => {
      const { query } = createTestClient(server);
      const page = 7;
      const pageSize = 18;
      const res = await query({
        query: GET_SPACE_CENTER_BY_CUSTOM_VARS,
        variables: {
          page,
          pageSize,
        },
      });
      expect(res.data.spaceCenters.nodes).toHaveLength(pageSize);
      expect(res.errors).not.toBeTruthy();
      expect(res).toMatchSnapshot();
    });

    it('fetch spacers with custom page which throw error (page < 1)', async () => {
      const { query } = createTestClient(server);
      const page = 0;
      const pageSize = 18;
      const res = await query({
        query: GET_SPACE_CENTER_BY_CUSTOM_VARS,
        variables: {
          page,
          pageSize,
        },
      });
      expect(res.errors).toBeTruthy();
      expect(res).toMatchSnapshot();
    });

    it('fetch spacers with custom pageSize which throw error (pageSize > 100)', async () => {
      const { query } = createTestClient(server);
      const page = 1;
      const pageSize = 0;
      const res = await query({
        query: GET_SPACE_CENTER_BY_CUSTOM_VARS,
        variables: {
          page,
          pageSize,
        },
      });
      expect(res.errors).toBeTruthy();
      expect(res).toMatchSnapshot();
    });

    it('fetch spacers with custom pageSize which throw error (pageSize > 100)', async () => {
      const { query } = createTestClient(server);
      const page = 1;
      const pageSize = 118;
      const res = await query({
        query: GET_SPACE_CENTER_BY_CUSTOM_VARS,
        variables: {
          page,
          pageSize,
        },
      });
      expect(res.errors).toBeTruthy();
      expect(res).toMatchSnapshot();
    });
  });
});
