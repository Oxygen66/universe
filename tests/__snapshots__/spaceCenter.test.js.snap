// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries spaceCenters fetch space centers  1`] = `
Object {
  "data": Object {
    "spaceCenters": Object {
      "nodes": Array [
        Object {
          "description": "name1",
          "id": "1",
          "latitude": -1,
          "longitude": -1,
          "name": "name1",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid1",
        },
        Object {
          "description": "name2",
          "id": "2",
          "latitude": -1,
          "longitude": -1,
          "name": "name2",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid2",
        },
        Object {
          "description": "name3",
          "id": "3",
          "latitude": -1,
          "longitude": -1,
          "name": "name3",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid3",
        },
        Object {
          "description": "name4",
          "id": "4",
          "latitude": -1,
          "longitude": -1,
          "name": "name4",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid4",
        },
        Object {
          "description": "name5",
          "id": "5",
          "latitude": -1,
          "longitude": -1,
          "name": "name5",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid5",
        },
        Object {
          "description": "name6",
          "id": "6",
          "latitude": -1,
          "longitude": -1,
          "name": "name6",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid6",
        },
        Object {
          "description": "name7",
          "id": "7",
          "latitude": -1,
          "longitude": -1,
          "name": "name7",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid7",
        },
        Object {
          "description": "name8",
          "id": "8",
          "latitude": -1,
          "longitude": -1,
          "name": "name8",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid8",
        },
        Object {
          "description": "name9",
          "id": "9",
          "latitude": -1,
          "longitude": -1,
          "name": "name9",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid9",
        },
        Object {
          "description": "name10",
          "id": "10",
          "latitude": -1,
          "longitude": -1,
          "name": "name10",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid10",
        },
      ],
      "pagination": Object {
        "page": 1,
        "pageSize": 10,
        "total": 1081,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries spaceCenters fetch space centers  2`] = `
Object {
  "data": Object {
    "spaceCenters": Object {
      "nodes": Array [
        Object {
          "description": "name1",
          "id": "1",
          "latitude": -1,
          "longitude": -1,
          "name": "name1",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid1",
        },
        Object {
          "description": "name2",
          "id": "2",
          "latitude": -1,
          "longitude": -1,
          "name": "name2",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid2",
        },
        Object {
          "description": "name3",
          "id": "3",
          "latitude": -1,
          "longitude": -1,
          "name": "name3",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid3",
        },
        Object {
          "description": "name4",
          "id": "4",
          "latitude": -1,
          "longitude": -1,
          "name": "name4",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid4",
        },
        Object {
          "description": "name5",
          "id": "5",
          "latitude": -1,
          "longitude": -1,
          "name": "name5",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid5",
        },
        Object {
          "description": "name6",
          "id": "6",
          "latitude": -1,
          "longitude": -1,
          "name": "name6",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid6",
        },
        Object {
          "description": "name7",
          "id": "7",
          "latitude": -1,
          "longitude": -1,
          "name": "name7",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid7",
        },
        Object {
          "description": "name8",
          "id": "8",
          "latitude": -1,
          "longitude": -1,
          "name": "name8",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid8",
        },
        Object {
          "description": "name9",
          "id": "9",
          "latitude": -1,
          "longitude": -1,
          "name": "name9",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid9",
        },
        Object {
          "description": "name10",
          "id": "10",
          "latitude": -1,
          "longitude": -1,
          "name": "name10",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid10",
        },
      ],
      "pagination": Object {
        "page": 1,
        "pageSize": 10,
        "total": 1081,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries spaceCenters fetch spacers with custom page which throw error (page < 1) 1`] = `
Object {
  "data": Object {
    "spaceCenters": null,
  },
  "errors": Array [
    [GraphQLError: Page must be greater than 1],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries spaceCenters fetch spacers with custom page, pageSize 1`] = `
Object {
  "data": Object {
    "spaceCenters": Object {
      "nodes": Array [
        Object {
          "description": "name109",
          "id": "109",
          "latitude": -1,
          "longitude": -1,
          "name": "name109",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid109",
        },
        Object {
          "description": "name110",
          "id": "110",
          "latitude": -1,
          "longitude": -1,
          "name": "name110",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid110",
        },
        Object {
          "description": "name111",
          "id": "111",
          "latitude": -1,
          "longitude": -1,
          "name": "name111",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid111",
        },
        Object {
          "description": "name112",
          "id": "112",
          "latitude": -1,
          "longitude": -1,
          "name": "name112",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid112",
        },
        Object {
          "description": "name113",
          "id": "113",
          "latitude": -1,
          "longitude": -1,
          "name": "name113",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid113",
        },
        Object {
          "description": "name114",
          "id": "114",
          "latitude": -1,
          "longitude": -1,
          "name": "name114",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid114",
        },
        Object {
          "description": "name115",
          "id": "115",
          "latitude": -1,
          "longitude": -1,
          "name": "name115",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid115",
        },
        Object {
          "description": "name116",
          "id": "116",
          "latitude": -1,
          "longitude": -1,
          "name": "name116",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid116",
        },
        Object {
          "description": "name117",
          "id": "117",
          "latitude": -1,
          "longitude": -1,
          "name": "name117",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid117",
        },
        Object {
          "description": "name118",
          "id": "118",
          "latitude": -1,
          "longitude": -1,
          "name": "name118",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid118",
        },
        Object {
          "description": "name119",
          "id": "119",
          "latitude": -1,
          "longitude": -1,
          "name": "name119",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid119",
        },
        Object {
          "description": "name120",
          "id": "120",
          "latitude": -1,
          "longitude": -1,
          "name": "name120",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid120",
        },
        Object {
          "description": "name121",
          "id": "121",
          "latitude": -1,
          "longitude": -1,
          "name": "name121",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid121",
        },
        Object {
          "description": "name122",
          "id": "122",
          "latitude": -1,
          "longitude": -1,
          "name": "name122",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid122",
        },
        Object {
          "description": "name123",
          "id": "123",
          "latitude": -1,
          "longitude": -1,
          "name": "name123",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid123",
        },
        Object {
          "description": "name124",
          "id": "124",
          "latitude": -1,
          "longitude": -1,
          "name": "name124",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid124",
        },
        Object {
          "description": "name125",
          "id": "125",
          "latitude": -1,
          "longitude": -1,
          "name": "name125",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid125",
        },
        Object {
          "description": "name126",
          "id": "126",
          "latitude": -1,
          "longitude": -1,
          "name": "name126",
          "planet": Object {
            "code": "COD",
            "id": "1",
            "name": "test1",
          },
          "uid": "uid126",
        },
      ],
      "pagination": Object {
        "page": 7,
        "pageSize": 18,
        "total": 1081,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries spaceCenters fetch spacers with custom pageSize which throw error (pageSize > 100) 1`] = `
Object {
  "data": Object {
    "spaceCenters": null,
  },
  "errors": Array [
    [GraphQLError: PageSize must be between 1 and 100],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries spaceCenters fetch spacers with custom pageSize which throw error (pageSize > 100) 2`] = `
Object {
  "data": Object {
    "spaceCenters": null,
  },
  "errors": Array [
    [GraphQLError: PageSize must be between 1 and 100],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
